File: 프로그래머스/2/118667. 두 큐 합 같게 만들기/두 큐 합 같게 만들기.py
from collections import deque

def solution(queue1, queue2):
    answer = -1
    total_sum = sum(queue1) + sum(queue2)
    target_sum = total_sum / 2

    if total_sum % 2 != 0:
        return -1

    current_q1 = deque(queue1)
    current_q2 = deque(queue2)
    current_sum = sum(current_q1)
    
    operations_count = 0
    
    if max(current_q1) > target_sum or max(current_q2) > target_sum:
        return -1
    
    while current_q1 and current_q2:
        if current_sum > target_sum:
            current_sum -= current_q1.popleft()
            operations_count += 1   
        elif current_sum < target_sum:
            element = current_q2.popleft()
            current_q1.append(element)
            current_sum += element
            operations_count += 1 
        elif current_sum == target_sum:
            return operations_count
    
    return answer

File: 프로그래머스/2/118667. 두 큐 합 같게 만들기/두 큐 합 같게 만들기.java
import java.util.*;

class Solution {
    public int solution(int[] queue1, int[] queue2) {
        Queue<Integer> currentQ1 = new LinkedList<>();
        Queue<Integer> currentQ2 = new LinkedList<>();
        
        long sumQ1 = 0;
        long sumQ2 = 0;
        
        for (int num : queue1) {
            currentQ1.offer(num);
            sumQ1 += num;
        }
        for (int num : queue2) {
            currentQ2.offer(num);
            sumQ2 += num;
        }
        
        int answer = -1;
        long totalSum = sumQ1 + sumQ2;
        long targetSum = totalSum / 2;

        if (totalSum % 2 != 0) {
            return -1;
        }
        
        long currentSum = sumQ1;

        int operationsCount = 0;

        if (max(currentQ1) > targetSum || max(currentQ2) > targetSum) {
            return -1;
        }

        while (!currentQ1.isEmpty() && !currentQ2.isEmpty()) {
            if (currentSum > targetSum) {
                currentSum -= currentQ1.poll();
                operationsCount++;
            } else if (currentSum < targetSum) {
                int element = currentQ2.poll();
                currentQ1.offer(element);
                currentSum += element;
                operationsCount++;
            } else if (currentSum == targetSum) {
                return operationsCount;
            }
        }

        return -1;
    }

    private static int max(Queue<Integer> queue) {
        int max = Integer.MIN_VALUE;
        for (int num : queue) {
            max = Math.max(max, num);
        }
        return max;
    }
}